test/general_test.py

```diff
@@ -0,0 +1,295 @@
+import os
+import subprocess
+import pytest
+import requests_unixsocket
+import socket
+import smtplib
+import poplib
+import ssl
+import base64
+import urllib.parse
+from requests.adapters import HTTPAdapter
+from urllib3.connection import HTTPConnection
+from urllib3.connectionpool import HTTPConnectionPool
+from urllib3.util.ssl_ import create_urllib3_context
+from email.message import EmailMessage
+from pathlib import Path
+
+CERT_PATH = "test/artifacts/ca_cert.pem"
+REQ_OPTS = {
+    "verify": CERT_PATH,
+    "timeout": 10
+}
+HTTPS_SOCKET_PATH = "socks/https.sock"
+POP3S_SOCKET_PATH = "./socks/pop3s.sock"
+SMTPS_SOCKET_PATH = "./socks/smtps.sock"
+REGISTER_PASS = ""
+
+
+class SSLUnixSocketConnection(HTTPConnection):
+    """Custom HTTPConnection that wraps a Unix socket with SSL."""
+
+    def __init__(self, unix_socket_path, **kwargs):
+        super().__init__("localhost", **kwargs)  # Dummy host (not used)
+        self.unix_socket_path = unix_socket_path
+        self.ssl_context = create_urllib3_context()
+        self.ssl_context.load_verify_locations(CERT_PATH)
+
+    def _new_conn(self):
+        """Create a new wrapped SSL connection over a Unix domain socket."""
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(self.unix_socket_path)
+        return self.ssl_context.wrap_socket(sock, server_hostname="localhost")
+
+
+class SSLUnixSocketConnectionPool(HTTPConnectionPool):
+    """Custom ConnectionPool for Unix socket wrapped with SSL."""
+
+    def __init__(self, socket_path, **kwargs):
+        super().__init__("localhost", **kwargs)  # Dummy host (not used)
+        self.socket_path = socket_path
+
+    def _new_conn(self):
+        return SSLUnixSocketConnection(self.socket_path)
+
+
+class SSLUnixSocketAdapter(HTTPAdapter):
+    """Transport adapter to route requests over a Unix socket with SSL."""
+
+    def __init__(self, socket_path, **kwargs):
+        self.socket_path = socket_path
+        super().__init__(**kwargs)
+
+    def get_connection(self, url, proxies=None):
+        return SSLUnixSocketConnectionPool(self.socket_path)
+
+
+class UnixPOP3(poplib.POP3):
+    def __init__(self, socket_path, certfile, timeout=30):
+        self.timeout = timeout
+        # Create SSL context and load certificate
+        ssl_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
+        ssl_context.load_verify_locations(certfile)
+
+        # Create and connect Unix domain socket
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(socket_path)
+        # Wrap with SSL
+        self.sock = ssl_context.wrap_socket(sock, server_hostname='localhost.localdomain')
+        self.file = self.sock.makefile('rb')
+        self._debugging = 0
+
+
+def singularity_pop_login(user, pass_):
+    mailbox = UnixPOP3(POP3S_SOCKET_PATH, CERT_PATH)
+    mailbox.user(user)
+    mailbox.pass_(pass_)
+    return mailbox
```

Comment:
maybe I am insane, but like can't we just use [pycurl](https://pypi.org/project/pycurl/) to actually one to one map the curl commands we were using instead of contorting python's request library to do things it does not want to?